src/main/java/org/truffleruby/core/hash/library/CompactHashStore.java

```diff
@@ -0,0 +1,506 @@
+
+import org.truffleruby.RubyContext;
+import org.truffleruby.RubyLanguage;
+import org.truffleruby.collections.PEBiFunction;
+import org.truffleruby.core.array.ArrayHelpers;
+import org.truffleruby.core.array.RubyArray;
+import org.truffleruby.core.hash.CompareHashKeysNode;
+import org.truffleruby.core.hash.FreezeHashKeyIfNeededNode;
+import org.truffleruby.core.hash.HashLiteralNode;
+import org.truffleruby.core.hash.HashingNodes;
+import org.truffleruby.core.hash.RubyHash;
+import org.truffleruby.language.RubyBaseNode;
+import org.truffleruby.language.RubyNode;
+import org.truffleruby.language.objects.shared.PropagateSharingNode;
+
+import com.oracle.truffle.api.CompilerDirectives;
+import com.oracle.truffle.api.dsl.Bind;
+import com.oracle.truffle.api.dsl.Cached;
+import com.oracle.truffle.api.dsl.Cached.Shared;
+import com.oracle.truffle.api.dsl.GenerateUncached;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.frame.Frame;
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.library.CachedLibrary;
+import com.oracle.truffle.api.library.ExportLibrary;
+import com.oracle.truffle.api.library.ExportMessage;
+import com.oracle.truffle.api.nodes.ExplodeLoop;
+import com.oracle.truffle.api.nodes.Node;
+import com.oracle.truffle.api.profiles.InlinedConditionProfile;
+
+/* The Compact hash strategy from Hash Maps That Dont Hate You
+ * See https://blog.toit.io/hash-maps-that-dont-hate-you-1a96150b492a for more details (archived at
+ * https://archive.ph/wip/sucRY) */
+@ExportLibrary(value = HashStoreLibrary.class)
+@GenerateUncached
+public class CompactHashStore {
+    // Big Picture :
+    //  (1) A Compact Hash's big idea is storing all key-values in a contiguous array (the KV array) in insertion order
+    //      This enables us to do sequential iteration in insertion order, which is what we want
+    //      But it doesn't allow constant-time lookups
+    //  -----------------------------------------------
+    //  (2) Thus, a Compact Hash also stores the position of every key in the KV array in another array called the index
+    //      That position is stored at a location that depends on the respective key's hash
+    //      Thus given a key we can always quickly find where it's (and we also store the hash itself in the index)
+    //  -----------------------------------------------
+    //  (3) Therefore, Compact Hashes give you both insertion-order iteration as well as the usual const-time guarantees
+    //  ---------------------------------------------------------------------------------------------------------------
+
+    /* We view the index array as consisting of "Slots", each slot is 2 array positions. index[0] and index[1] is one
+     * slot, index[2] and index[3] is another,... The first position of a slot holds the hash of a key and the second
+     * holds an offsetted index into the KV store (index + 1) where the key is stored. The reason we store an offsetted
```

Comment:
> Do we change to 1-based indexing

Yes, exactly. Every kv position is incremented by 1 before storage in the index array, and then decremented again on retrieval. 

(All kv positions stored in the index array would be positions for keys, and since keys are always stored at even positions in KV storage {0,2,...}, that means all kv positions stored in the index array are odd numbers {1,3,...}. That's an expensive constraint to check in .verify(), so I don't. But it should be kept in mind as a debugging redflag.)  

> A negative value would look like a better tombstone to me

Same here, 1-shifting was Benoit's idea. Great idea to avoid initializing the array manually, but a rich source of potential bugs to be very careful around !

Reply:
>  Or is the initial entry in the KV store not a validate storage candidate

Nope, all entries in the KV is fair game. 

(Deletion from an entry results in null-marking that entry so we don't stumble on it later, and then that entry is never used again, but that's a different issue. Still bad.)